ProvisionRestResult implementation

tiagralee opened the pull request:
Custom handler implementations in some instances require access to
more information than downstream service calls currently permit. This
can include the status code, headers or body in the instance that a
non-successful response was returned.

Custom code can ProvisionRestResult and retrive the RestResult after the downstream call
Generated code will set the HTTP result as the RestResult

close #105 #159

andrewemeryanz commented on common/corerequestcontext.go:
```diff
@@ -171,3 +177,21 @@ func (t *tempRoundtripper) RoundTrip(r *http.Request) (*http.Response, error) {
+
+// Provision within the context the ability to retrieve the result of a rest request
+func ProvisionRestResult(ctx context.Context) context.Context {
+	// insert an object in the context
```
Remove this comment:
```go
// insert an object in the context
```

andrewemeryanz commented on restlib/restlib.go:
```diff
@@ -193,3 +193,29 @@ func SetHeaders(w http.ResponseWriter, headerMap http.Header) {
+
+// OnRestResult method will be called from tests as we don't expose the HTTPResult itself
+func OnRestResult(ctx context.Context, result *common.RestResult, err error) {
+	// find the object in the context (if available) and set the result
```
Remove this comment:
```go
// find the object in the context (if available) and set the result
```

andrewemeryanz commented on restlib/restlib.go:
```diff
@@ -193,3 +193,29 @@ func SetHeaders(w http.ResponseWriter, headerMap http.Header) {
+
+// OnRestResultHTTPResult is called from generated code when an HTTP result is retrieved.
+func OnRestResultHTTPResult(ctx context.Context, result *HTTPResult, err error) {
+	if result == nil || result.HTTPResponse == nil {
```
This logic isn't correct. If we only have an error then we want to propagate that also:
```go
var restResult *common.RestResult = nil
if result != nil && result.HTTPResponse != nil {
    restResult = ...
}
OnRestResult(ctx, restResult, err)
```

andrewemeryanz reviewed (changes requested):
Great. Couple of minor changes.

Commit:
ProvisionRestResult implementation

Custom handler implementations in some instances require access to
more information than downstream service calls currently permit. This
can include the status code, headers or body in the instance that a
non-successful response was returned.

Custom code can ProvisionRestResult and retrive the RestResult after the downstream call
Generated code will set the HTTP result as the RestResult

close #105 #159

andrewemeryanz reviewed (approved):
Perfect.
